src/abi/anchor.rs

```diff
@@ -17,3 +28,375 @@ pub fn discriminator(namespace: &'static str, name: &str) -> Vec<u8> {
+            for (item_no, item) in def.fields.iter().enumerate() {
+                let name = if item.id.is_none() {
+                    format!("field_{}", item_no)
+                } else {
+                    item.name_as_str().to_string()
+                };
+
+                fields.push(IdlEventField {
+                    name,
+                    ty: type_manager.convert(&item.ty),
+                    index: item.indexed,
+                });
+            }
+
+            events.push(IdlEvent {
+                name: def.name.clone(),
+                fields,
+            });
+        }
+
+        Some(events)
+    }
+}
+
+/// Generate the IDL instructions for a contract.
+fn idl_instructions(
+    contract_no: usize,
+    contract: &Contract,
+    type_manager: &mut TypeManager,
+    ns: &Namespace,
+) -> Vec<IdlInstruction> {
+    let mut instructions: Vec<IdlInstruction> = Vec::new();
+
+    if !contract.have_constructor(ns) {
+        instructions.push(IdlInstruction {
+            name: "new".to_string(),
+            docs: None,
+            accounts: vec![IdlAccountItem::IdlAccount(IdlAccount {
+                name: "data_account".to_string(),
+                is_mut: true,
+                is_signer: false,
+                is_optional: Some(false),
+                docs: None,
+                pda: None,
+                relations: vec![],
+            })],
+            args: vec![],
+            returns: None,
+        })
+    }
+
+    for func_no in contract.all_functions.keys() {
+        if !ns.functions[*func_no].is_public() {
+            continue;
+        }
+
+        let func = &ns.functions[*func_no];
+        let tags = idl_docs(&func.tags);
+
+        let accounts = match &func.mutability {
+            Mutability::Pure(_) => {
+                vec![]
+            }
+            Mutability::View(_) => {
+                vec![IdlAccountItem::IdlAccount(IdlAccount {
+                    name: "data_account".to_string(),
+                    is_mut: false,
+                    is_signer: false,
+                    is_optional: Some(false),
+                    docs: None,
+                    pda: None,
+                    relations: vec![],
+                })]
+            }
+            _ => {
+                vec![IdlAccountItem::IdlAccount(IdlAccount {
+                    name: "data_account".to_string(),
+                    is_mut: true,
+                    is_signer: false,
+                    is_optional: Some(false),
+                    docs: None,
+                    pda: None,
+                    relations: vec![],
+                })]
+            }
+        };
+
+        let mut args: Vec<IdlField> = Vec::with_capacity(func.params.len());
+        for (item_no, item) in func.params.iter().enumerate() {
+            let name = if item.id.is_none() || item.id.as_ref().unwrap().name.is_empty() {
+                format!("arg_{}", item_no)
```

Comment:
This is a parameter, right? Also no check for dups

Reply:
Function arguments must be named. I should add a comment about that.

Reply:
Function argument don't have to be named